tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,279 @@
+  }
+
+  #bring up the SR-IOV device
+  foreach $pf (@pfs) {
+    my $nic = script_output "ls -l /sys/class/net |grep $pf | awk '{print \$9}'";
+    assert_script_run "echo \"BOOTPROTO='dhcp'\nSTARTMODE='manual'\" > /etc/sysconfig/network/ifcfg-$nic";
+    assert_script_run("ifup $nic", 60);   #15s last time
+  }
+  
+  my $vfs = script_output "lspci | grep Ethernet | grep \"Virtual Function\" | cut -d ' ' -f1";
+  my @vfs = split("\n", $vfs);
+
+}
+
+
+#get vf device id in libvirt by BDF
+sub get_device_id(){
+  my $device_bdf = shift;
+
+  #change to device id in libvirt
+  $device_bdf =~ s/[:\.]/_/g;
+  #once get nothing while device is listed after a moment
+  my $device_id = script_output "virsh nodedev-list | grep $device_bdf";  
+  return 1 if ! $device_id;
+  $device_id;
+
+}
+
+
+#plugin a device to vm via libvirt
+sub plugin_device_kvm() {
+  my ($vm, $host_device_id) = @_;
+
+  #get neccessary device config from host
+  my $host_device_xml = script_output "virsh nodedev-dumpxml $host_device_id";
+  $host_device_xml =~ /\<domain\>(\w+)\<\/domain\>.*\<bus\>(\w+)\<\/bus\>.*\<slot\>(\w+)\<\/slot\>.*\<function\>(\w+)\<\/function\>/s;
+  my ($dev_domain, $dev_bus, $dev_slot, $dev_func) = ($1, $2, $3, $4);
+  
+  #'virsh attach-interface' can plug the network device to guest with the BDF in commandline without a seperate device xml file
+  #but the SUSE document use 'virsh attach-device', so the test follows it
+  #create the device xml to passthrough to vm
+  assert_script_run("echo \"<interface type='hostdev'>\n  <source>\n    <address type='pci' domain='$dev_domain' bus='$dev_bus' slot='$dev_slot' function='$dev_func'/>\n  </source>\n</interface>\" > $host_device_id.xml", 60);
+
+  #attach device to vm
+  assert_script_run "virsh -d 1 attach-device $vm $host_device_id.xml --persistent";
+  save_screenshot;
+
+  #check the vm xml
+  script_run "virsh dumpxml $vm | grep -A8 interface";
+
+  #get mac address
+  #we assume that only one SR-IOV vf is plugged in guest
+  my %nic;
+  $nic{mac} = script_output "virsh domiflist $vm | grep 'hostdev' | grep -oE \"[[:xdigit:]]{2}(:[[:xdigit:]]{2}){5}\"";
+
+  #get the network device's BDF
+  #get BDF this way(not from vm config xml) to check the device meanwhile
+  #should consider get it from /sys/class/net/ via mac address ???
+  $nic{bdf} = script_output "ssh root\@$vm \"lspci | grep Eth | grep 'Virtual Function' | cut -d ' ' -f1\"";
+  save_screenshot;
+  print "julies: \$nic{bdf}=$nic{bdf}\n";
+  script_run "ssh root\@$vm \"lspci -vvv -s $nic{bdf}\"";
+  save_screenshot;  
+
+  return %nic;
+
+}
+
+
+#unplug the vf device from guest
+sub unplug_vf_from_vm {
+  my ($vm, $mac) = @_;
+
+  #create a device xml
+  my $vf_xml = script_output "virsh dumpxml $vm | sed -n '/<interface.*hostdev/,/\\\/interface>/p'";
+  return 1 if (! $vf_xml =~ /$mac/);
+  my $vf_xml_file = "vf_in_vm.xml";
+  assert_script_run "echo \"$vf_xml\" > $vf_xml_file";
+  script_run "cat $vf_xml_file";  #julie debug
```

Comment:
imho, if this created device xml file was useful, suggestion to use upload_log () here.

Reply:
good suggestion. this xml file and another xml file should be uploaded as logs. I overlooked them.